astropy/cosmology/parameter.py

```diff
@@ -210,8 +214,93 @@ def register(fvalidate):
 
     # -------------------------------------------
 
+    @property
+    def _init_arguments(self):
+        """Initialization arguments."""
+        kw = {"derived": self.derived,
+              "unit": self.unit,
+              "equivalencies": self.equivalencies,
+              "fmt": self.format_spec,
+              "doc": self.__doc__}
+        # validator is always turned into a function, but for ``repr`` it's nice
+        # to know if it was originally a string.
+        fvalidate = (self._validate_repr if self._validate_repr in self._registry_validators
+                     else self.fvalidate)
+        kw["fvalidate"] = fvalidate
+        return kw
+
+    def clone(self, **kw):
+        """Clone this `Parameter`, changing any constructor argument.
+
+        Parameters
+        ----------
+        **kw
+            Passed to constructor. The current values, eg. ``fvalidate`` are
+            used as the default values, so an empty ``**kw`` is an exact copy.
+
+        Examples
+        --------
+        >>> p = Parameter()
+        >>> p
+        Parameter(derived=False, unit=None, equivalencies=[],
+                  fvalidate='default', fmt='.3g', doc=None)
+
+        >>> p.clone(unit="km")
+        Parameter(derived=False, unit=Unit("km"), equivalencies=[],
+                  fvalidate='default', fmt='.3g', doc=None)
+        """
+        # start with defaults, update from kw.
+        kwargs = {**self._init_arguments, **kw}
+        # all initialization failures, like incorrect input are handled by init
+        cloned = type(self)(**kwargs)
+        # transfer over the __set_name__ stuff. If `clone` is used to make a
+        # new descriptor, __set_name__ will be called again, overwriting this.
+        cloned._attr_name = self._attr_name
+        cloned._attr_name_private = self._attr_name_private
+
+        return cloned
+
+    def __eq__(self, other):
+        """Check Parameter equality. Only equal to other Parameter objects.
+
+        Returns
+        -------
+        NotImplemented or True
+            `True` if equal, `NotImplemented` otherwise. This allows `other` to
+            be check for equality with ``other.__eq__``.
+
+        Examples
+        --------
+        >>> p1, p2 = Parameter(unit="km"), Parameter(unit="km")
+        >>> p1 == p2
+        True
+
+        >>> p3 = Parameter(unit="km / s")
+        >>> p3 == p1
+        False
+
+        >>> p1 != 2
+        True
+        """
+        if not isinstance(other, Parameter):
+            return NotImplemented
+        # check equality on all @property
+        for n in self._properties:  # pass to ``other.__eq__`` if fails check
```

Comment:
It seems `_properties` and `_init_arguments` serve relatively similar purposes. Could one do:
```
return self._init_arguments == other._init_arguments
```
Note that here we *know* other is a Parameter, so there we can count on `_init_arguments` to be present (and any subclass that overrides its meaning or needs more information will be called before us)

Reply:
``return self._init_arguments == other._init_arguments`` doesn't quite work.
``_init_arguments`` reverts a function back to a string, if possible. So
``p1 = Parameter(fvalidate="scalar")`` and ``p2 = Parameter(fvalidate=p1.fvalidate)`` would be different because the latter cannot be reverted to a string.
I've fixed this by making 

``_init_arguments = property(_get_init_arguments)`` and adding `` _get_init_arguments``, which can return either the string or function form of ``fvalidate``.